File: src/test/java/com/pubnub/api/managers/BasePathManagerTest.java
package com.pubnub.api.managers;

import com.pubnub.api.PNConfiguration;
import com.pubnub.api.PubNubException;
import com.pubnub.api.UserId;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.io.IOException;
import java.util.UUID;

public class BasePathManagerTest {

    private PNConfiguration pnConfiguration;

    @Before
    public void beforeEach() throws IOException, PubNubException {
        pnConfiguration =  new PNConfiguration(new UserId("pn-" + UUID.randomUUID()));
    }

    @Test
    public void stdOriginNotSecure() {
        pnConfiguration.setSecure(false);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("http://ps.pndsn.com", basePathManager.getBasePath());
    }

    @Test
    public void stdOriginSecure() {
        pnConfiguration.setSecure(true);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("https://ps.pndsn.com", basePathManager.getBasePath());
    }

    @Test
    public void customOriginNotSecure() {
        pnConfiguration.setOrigin("custom.origin.com");
        pnConfiguration.setSecure(false);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("http://custom.origin.com", basePathManager.getBasePath());
    }

    @Test
    public void customOriginSecure() {
        pnConfiguration.setOrigin("custom.origin.com");
        pnConfiguration.setSecure(true);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("https://custom.origin.com", basePathManager.getBasePath());
    }

    @SuppressWarnings("deprecation")
    @Test
    public void customOriginNotSecureWithCacheBusting() {
        pnConfiguration.setOrigin("custom.origin.com");
        pnConfiguration.setCacheBusting(true);
        pnConfiguration.setSecure(false);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("http://custom.origin.com", basePathManager.getBasePath());
    }

    @SuppressWarnings("deprecation")
    @Test
    public void customOriginSecureWithCacheBusting() {
        pnConfiguration.setOrigin("custom.origin.com");
        pnConfiguration.setSecure(true);
        pnConfiguration.setCacheBusting(true);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("https://custom.origin.com", basePathManager.getBasePath());
    }

    @SuppressWarnings("deprecation")
    @Test
    public void cacheBustingNotSecure() {
        pnConfiguration.setCacheBusting(true);
        pnConfiguration.setSecure(false);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("http://ps1.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps2.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps3.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps4.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps5.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps6.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps7.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps8.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps9.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps10.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps11.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps12.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps13.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps14.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps15.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps16.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps17.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps18.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps19.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps20.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("http://ps1.pndsn.com", basePathManager.getBasePath());
    }

    @SuppressWarnings("deprecation")
    @Test
    public void cacheBustingSecure() {
        pnConfiguration.setCacheBusting(true);
        BasePathManager basePathManager = new BasePathManager(pnConfiguration);
        Assert.assertEquals("https://ps1.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps2.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps3.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps4.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps5.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps6.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps7.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps8.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps9.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps10.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps11.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps12.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps13.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps14.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps15.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps16.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps17.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps18.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps19.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps20.pndsn.com", basePathManager.getBasePath());
        Assert.assertEquals("https://ps1.pndsn.com", basePathManager.getBasePath());
    }

}

File: src/test/kotlin/com/pubnub/api/legacy/managers/BasePathManagerTest.kt
package com.pubnub.api.legacy.managers

import com.pubnub.api.legacy.BaseTest
import com.pubnub.api.managers.BasePathManager
import org.junit.Assert.assertEquals
import org.junit.Test

class BasePathManagerTest : BaseTest() {

    override fun onBefore() {
        clearConfiguration()
    }

    @Test
    fun stdOriginNotSecure() {
        config.secure = false
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("http://ps.pndsn.com", basePathManager.basePath())
    }

    @Test
    fun stdOriginSecure() {
        config.secure = true
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("https://ps.pndsn.com", basePathManager.basePath())
    }

    @Test
    fun customOriginNotSecure() {
        config.origin = "custom.origin.com"
        config.secure = false
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("http://custom.origin.com", basePathManager.basePath())
    }

    @Test
    fun customOriginSecure() {
        config.origin = "custom.origin.com"
        config.secure = true
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("https://custom.origin.com", basePathManager.basePath())
    }

    @Test
    fun customOriginNotSecureWithCacheBusting() {
        config.origin = "custom.origin.com"
        config.cacheBusting = true
        config.secure = false
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("http://custom.origin.com", basePathManager.basePath())
    }

    @Test
    fun customOriginSecureWithCacheBusting() {
        config.origin = "custom.origin.com"
        config.secure = true
        config.cacheBusting = true
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("https://custom.origin.com", basePathManager.basePath())
    }

    @Test
    fun cacheBustingNotSecure() {
        config.cacheBusting = true
        config.secure = false
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("http://ps1.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps2.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps3.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps4.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps5.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps6.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps7.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps8.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps9.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps10.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps11.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps12.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps13.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps14.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps15.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps16.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps17.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps18.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps19.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps20.pndsn.com", basePathManager.basePath())
        assertEquals("http://ps1.pndsn.com", basePathManager.basePath())
    }

    @Test
    fun cacheBustingSecure() {
        config.cacheBusting = true
        initPubNub()
        val basePathManager = BasePathManager(config)
        assertEquals("https://ps1.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps2.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps3.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps4.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps5.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps6.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps7.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps8.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps9.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps10.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps11.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps12.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps13.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps14.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps15.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps16.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps17.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps18.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps19.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps20.pndsn.com", basePathManager.basePath())
        assertEquals("https://ps1.pndsn.com", basePathManager.basePath())
    }
}
